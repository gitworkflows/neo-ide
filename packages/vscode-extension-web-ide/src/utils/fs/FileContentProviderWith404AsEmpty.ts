import type { FileContentProvider } from '@gitlab/web-ide-fs';
import { is404Error } from '@gitlab/gitlab-api-client';

/**
 * Decorator for FileContentProvider that returns 404 responses as empty
 */
export class FileContentProviderWith404AsEmpty implements FileContentProvider {
  readonly #base: FileContentProvider;

  constructor(base: FileContentProvider) {
    this.#base = base;
  }

  async getContent(path: string): Promise<Uint8Array> {
    try {
      // note: ESLint doesn't like it if I just return here...
      const value = await this.#base.getContent(path);

      return value;
    } catch (e: unknown) {
      if (is404Error(e)) {
        return new Uint8Array(0);
      }

      throw e;
    }
  }
}
