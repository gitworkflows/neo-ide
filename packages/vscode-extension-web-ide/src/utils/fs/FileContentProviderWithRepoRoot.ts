import type { FileContentProvider } from '@gitlab/web-ide-fs';
import { stripPathRoot } from '../stripPathRoot';

/**
 * Decorator for FileContentProvider that strips repo root from calls to getContent
 */
export class FileContentProviderWithRepoRoot implements FileContentProvider {
  readonly #base: FileContentProvider;

  readonly #repoRoot: string;

  constructor(base: FileContentProvider, repoRoot: string) {
    this.#base = base;
    this.#repoRoot = repoRoot;
  }

  getContent(path: string): Promise<Uint8Array> {
    return this.#base.getContent(stripPathRoot(path, this.#repoRoot));
  }
}
