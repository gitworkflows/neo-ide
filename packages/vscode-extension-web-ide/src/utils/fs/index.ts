export * from './FileContentProviderWith404AsEmpty';
export * from './FileContentProviderWithRepoRoot';
export * from './touchFile';
export * from './tryStat';
