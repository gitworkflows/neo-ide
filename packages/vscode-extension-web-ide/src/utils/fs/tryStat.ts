import type * as vscode from 'vscode';

export const tryStat = async (fs: vscode.FileSystem, uri: vscode.Uri) =>
  fs.stat(uri).then(
    x => x,
    () => null,
  );
