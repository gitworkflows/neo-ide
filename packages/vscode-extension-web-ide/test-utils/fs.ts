import { FileSystem } from '@gitlab/web-ide-fs';

export const createFileSystemMock = (): jest.Mocked<FileSystem> => ({
  stat: jest.fn().mockResolvedValue({
    ctime: 0,
    mode: 0,
    mtime: 0,
    size: 0,
    type: 0,
  }),
  readdir: jest.fn().mockResolvedValue([]),
  readdirWithTypes: jest.fn().mockResolvedValue([]),
  mkdir: jest.fn(),
  readFile: jest.fn().mockResolvedValue(new Uint8Array()),
  writeFile: jest.fn(),
  rm: jest.fn(),
  rename: jest.fn(),
  lastModifiedTime: jest.fn().mockResolvedValue(0),
});
