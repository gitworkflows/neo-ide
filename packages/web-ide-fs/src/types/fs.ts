export enum FileType {
  Blob = 1,
  Tree = 2,
}

export interface FileStats {
  readonly mode: number;
  readonly ctime: number;
  readonly mtime: number;
  readonly size: number;
  readonly type: FileType;
}

export interface GitLsTreeEntry {
  readonly path: string;
  readonly mode: string;
}

/**
 * Responsible for providing a friendly read/write FileSystem
 *
 * We want clients of web-ide-fs to depend on this interface so
 * that we are less coupled to implementation details (like BrowserFS)
 * and the interactions are simpler.
 */
export interface FileSystem {
  stat(path: string): Promise<FileStats>;

  readdir(path: string): Promise<string[]>;

  readdirWithTypes(path: string): Promise<[string, FileType][]>;

  mkdir(path: string): Promise<void>;

  readFile(path: string): Promise<Uint8Array>;

  writeFile(path: string, data: Uint8Array): Promise<void>;

  // TODO: Remove opt: { recursive }. It isn't being used...
  rm(path: string, opt: { recursive: boolean }): Promise<void>;

  rename(oldPath: string, newPath: string): Promise<void>;

  // THOUGHT: We might want to abstract this out of this type...
  lastModifiedTime(): Promise<number>;
}

/**
 * Responsible for providing file content based on the given path
 */
export interface FileContentProvider {
  getContent(path: string): Promise<Uint8Array>;
}

export interface FileList {
  listAllBlobs(): Promise<string[]>;
}
