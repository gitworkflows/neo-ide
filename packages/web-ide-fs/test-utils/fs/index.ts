export * from './constants';
export * from './createDefaultOverlayFS';
export * as bfsUtils from './browserFSUtils';
export * from './stringToBuffer';
