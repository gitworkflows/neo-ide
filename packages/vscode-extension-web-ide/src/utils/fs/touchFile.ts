import type * as vscode from 'vscode';
import { splitParent } from '@gitlab/utils-path';
import { tryStat } from './tryStat';

export const touchFile = async (fs: vscode.FileSystem, uri: vscode.Uri): Promise<void> => {
  const stat = await tryStat(fs, uri);

  // If we already exist, there's nothing to do
  if (stat) {
    return;
  }

  const [parentPath] = splitParent(uri.path);

  if (parentPath) {
    await fs.createDirectory(uri.with({ path: parentPath }));
  }

  await fs.writeFile(uri, new Uint8Array(0));
};
