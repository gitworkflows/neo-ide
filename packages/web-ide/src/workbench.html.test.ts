import { readFile } from 'fs/promises';
import { JSDOM } from 'jsdom';
import { join } from 'path';

describe('workbench.html', () => {
  const WORKBENCH_WINDOW_ORIGIN = 'https://example.com';

  let workbenchWindow: Window;

  const createWorkbenchWindow = async (): Promise<Window> => {
    const workbenchHtml = await readFile(join(__dirname, '../assets/workbench.html'), 'utf-8');

    return new JSDOM(workbenchHtml, { url: WORKBENCH_WINDOW_ORIGIN }).window as unknown as Window;
  };

  beforeEach(async () => {
    workbenchWindow = await createWorkbenchWindow();
  });

  it('loads vscode-bootstrap module', async () => {
    const script = workbenchWindow.document.querySelector('script');

    expect(script?.getAttribute('type')).toBe('module');
    expect(script?.innerHTML.trim()).toMatchSnapshot();
  });

  it('loads vscode CSS', async () => {
    const link = workbenchWindow.document.querySelector('link');

    expect(link?.getAttribute('href')).toBe(`../vscode/out/vs/workbench/workbench.web.main.css`);
  });

  it('adds http-equiv meta with CSP policy', () => {
    const meta = workbenchWindow.document.querySelector('meta[http-equiv]');

    expect(meta?.getAttribute('content')).toBe(
      "script-src 'self' 'sha256-C5NYHanFzD/3S/E+C9Ga8zzHxTUaR7hyCqrbSjck17w=' 'wasm-unsafe-eval'; worker-src blob:",
    );
  });
});
