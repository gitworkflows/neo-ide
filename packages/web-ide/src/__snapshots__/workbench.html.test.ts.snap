// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`workbench.html loads vscode-bootstrap module 1`] = `
"import { main } from '../main.js';

      main();"
`;
